### ISSUE TITLE
Default device selection

### ISSUE THREAD

**Issue by reguly** [clang-ykt/clang#7]
When running over MPI, different processes should pick up different GPUs (when the devices are in exclusive mode). Without explicitly picking a device with omp_set_default_device(X);, the application seems to silently fail to pick one up, and I guess reverts to running on the CPU. Inserting omp_set_default_device(world_rank) does solve the problem, but in a HPC environment one should avoid doing that, and let the runtime pick an available GPU automatically.

[mpitest.txt](https://github.com/clang-ykt/clang/files/619992/mpitest.txt)
mpicc   -O3 -fopenmp=libomp -fopenmp-targets=nvptx64-nvidia-cuda mpi_test.c
mpirun -np 2 nvprof ./a.out
shows that only one GPU is actually running kernels

**Comment by arpith-jacob** [comment]
@reguly I think the preferred way to do this is for the MPI scheduler to set the environment variable $OMP_DEFAULT_DEVICE appropriately.  The offload library makes use of this variable.

**Comment by reguly** [comment]
When this is not set, would it not be possible to let the CUDA runtime pick a device automatically instead of forcing the choice of device 0 and failing? This is something most people in the CUDA community rely on, but of course this is not CUDA, and I am unsure what the conventions are.

**Comment by arpith-jacob** [comment]
I'm not sure I understand how the CUDA runtime knows to map a logical device number to a physical device in an MPI environment.  Can you point me to example code?

**Comment by reguly** [comment]
In general, I don't think there needs to be an exact mapping. In any environment, by default under CUDA, it will select the first available GPU for you, unless you force the choice. How exactly it does that, I do not know. With the runtime API, I have the following really simple piece of code that just loops over all devices and tries to select them, in effect reproducing that behaviour.

int deviceCount;
cudaGetDeviceCount(&deviceCount);
float *test;
  int gpu = -1;
  for (int i = 0; i < deviceCount; i++) {
    cudaError_t err = cudaSetDevice(i);
    if (err == cudaSuccess) {
      cudaError_t err = cudaMalloc((void **)&test, sizeof(float));
      if (err == cudaSuccess) {
        gpu = i;
        break;
      }
    }
  }
if (gpu == -1)
//we couldn't select any device...
else
cudaFree(test);

**Comment by arpith-jacob** [comment]
I'm not an expert on this so take my comment with a pinch of salt.  In general HPC users would like to bind a host thread in an MPI process and its offload device to the same NUMA domain.  Selecting a random device like you suggest will not be performant.

**Comment by reguly** [comment]
You are right that choice may not be performant, but it’s better than failing to pick up a GPU. Unfortunately many job schedulers currently do not set which GPU to pick up, but for the ones that do, you should look at that - i.e. CUDA_VISIBLE_DEVICES or OMP_DEFAULT_DEVICE. If these are not defined, then I think you should fall back to trying to pick up the first available device. This will keep it consistent with the way CUDA applications pick up devices.

**Comment by arpith-jacob** [comment]
Remember, OpenMP does not care if the offload target is a GPU or some other accelerator.  It does not assume a GPU from one particular vendor either.  I see no reason why an implementation of the offload library for OpenMP should look at vendor or accelerator specific variables.  The portable way is to use OMP_DEFAULT_DEVICE.

**Comment by reguly** [comment]
Okay, we'll just have to get used to setting it ourselves :)

**Comment by reguly** [comment]
Sorry, need to resurrect this:
CUDA_VISIBLE_DEVICES and OMP_DEFAULT_DEVICE do not work together.
CUDA_VISIBLE_DEVICES=0 OMP_DEFAULT_DEVICE=0 ./any_omp4code
crashes.
Due to reasons discussed previously, it will be the responsibility of the job scheduler to set OMP_DEFAULT_DEVICE correctly, however, for the same reasons it will want to set CUDA_VISIBLE_DEVICES as well, but the two will not work together.
